blob/service.go

```diff
@@ -0,0 +1,278 @@
+// and return its Proof.
+func (s *Service) GetProof(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Proof, error) {
+	_, proof, err := s.getByCommitment(ctx, height, nID, commitment)
+	if err != nil {
+		return nil, err
+	}
+	return proof, nil
+}
+
+// GetAll returns all found blobs for the given namespaces at the given height.
+func (s *Service) GetAll(ctx context.Context, height uint64, nIDs ...namespace.ID) ([]*Blob, error) {
+	header, err := s.getByHeight(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		blobs = make([]*Blob, 0)
+		ch    = make(chan []*Blob)
+	)
+
+	for _, nID := range nIDs {
+		log.Infow("requesting blobs", "height", height, "nID", nID.String())
+		go func(nID namespace.ID) {
+			blobs, err := s.getBlobs(ctx, nID, header.DAH)
+			if err != nil {
+				ch <- nil
+				return
+			}
+			ch <- blobs
+		}(nID)
+	}
+
+	for range nIDs {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case b := <-ch:
+			if b == nil {
+				continue
+			}
+			blobs = append(blobs, b...)
+		}
+	}
+
+	if len(blobs) == 0 {
+		return nil, errBlobNotFound
+	}
+	return blobs, nil
+}
+
+// Included verifies that the blob was included in a specific height.
+// To ensure that blob was included in a specific height, we need:
+// 1. verify the provided commitment by recomputing it;
+// 2. verify the provided Proof against subtree roots that were used in 1.;
+func (s *Service) Included(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	_ *Proof,
+	com Commitment,
+) (bool, error) {
+	// In the current implementation, LNs will have to download all shares to recompute the commitment.
+	// To achieve 1. we need to modify Proof structure and to store all subtree roots, that were
+	// involved in commitment creation and then call `merkle.HashFromByteSlices`(tendermint package).
+	// nmt.Proof is verifying share inclusion by recomputing row roots, so, theoretically, we can do
+	// the same but using subtree roots. For this case, we need an extra method in nmt.Proof
+	// that will perform all reconstructions,
+	// but we have to guarantee that all our stored subtree roots will be on the same height(e.g. one
+	// level above shares).
+	// TODO(@vgonkivs): rework the implementation to perform all verification without network requests.
+	_, _, err := s.getByCommitment(ctx, height, nID, com)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// getByCommitment retrieving DAH row by row, fetching shares and constructing blobs in order to
+// compare Commitments. Retrieving will be stopped once requested blob/proof will be found.
+func (s *Service) getByCommitment(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Blob, *Proof, error) {
+	log.Infow("requesting blob",
+		"height", height,
+		"nID", nID.String(),
+		"commitment", commitment.String())
```

Comment:
Agreed in sync to remove from logs, but later on change the stringer for nID and commitment to output something like base64. There are logs in other parts of the codebase that do nID.String(), so this will be in a separate PR

Reply:
removed